proxy/envoy/watcher.go

```diff
@@ -97,13 +103,45 @@ func (w *watcher) Reload() {
 	w.agent.ScheduleConfigUpdate(config)
 }
 
-// watchCerts watches a certificate directory and calls the provided
+type watchFileEventsFn func(ctx context.Context, wch <-chan *fsnotify.FileEvent,
+	maxDelay time.Duration, notifyFn func())
+
+// watchFileEvents watches for changes on a channel and notifies via notifyFn().
+// The function batches changes so that related changes are processed together.
+// The function ensures that notifyFn() is called no more than one time per maxDelay.
+// The function does not return until the the context is cancelled.
+func watchFileEvents(ctx context.Context, wch <-chan *fsnotify.FileEvent, maxDelay time.Duration, notifyFn func()) {
+	// timer and channel for managing maxDelay.
+	var timeChan <-chan time.Time
+	ticker := time.NewTicker(maxDelay)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case ev := <-wch:
+			glog.Infof("watchFileEvents: %s", ev.String())
+			timeChan = ticker.C
+		case <-timeChan:
+			// reset timer
+			timeChan = nil
```

Comment:
i think having a boolean that is set to true in wch block and then reset in timeChan block is cleaner. setting to `nil` makes me suspicious what happens with reading from nil channel.

Reply:
A nil channel is never ready. So it is never selected.

Reply:
reading from nil channel is idomatic go way of blocking